Add function to extract DH/ECDH from connection

FWinterborn opened the pull request:
Added a function to extract the ECDH and DH information from the OpenSSL connection. Modified test server to allow curve to be specified

FWinterborn committed:
Add function to extract DH/ECDH from connection

nabla-c0d3 commented on nassl/ssl_client.py:
```diff
@@ -298,6 +298,9 @@ def get_current_cipher_name(self) -> str:
         return self._ssl.get_cipher_bits()
 
+    def get_dh_info(self) -> dict:
+        return self._ssl.get_dh_info()
```
Would it be possible here to return a dataclass (or one of multiple possible dataclasses) instead of a dict, in order to document the fields (type, curve, etc.)? Enum fields are especially useful for that. In the current implementation knowing what gets returned here requires looking at the C code.

Example of a dataclass here: https://github.com/nabla-c0d3/sslyze/blob/master/sslyze/plugins/certificate_info_plugin.py#L906

nabla-c0d3 commented on nassl/_nassl/nassl_SSL.c:
```diff
@@ -813,6 +814,191 @@ static PyObject* nassl_SSL_get0_verified_chain(nassl_SSL_Object *self, PyObject
+        // Convert the public key to hex format
+        char* pos = public_key_hex;
+        for(int i = 0; i < len; i++){
+            pos += sprintf(pos, "%02X", public_key_bytes[i]);
```
Could it be converted to bytes instead? So it's directly "usable" from Python.

nabla-c0d3 commented on nassl/_nassl/nassl_SSL.c:
```diff
@@ -813,6 +814,191 @@ static PyObject* nassl_SSL_get0_verified_chain(nassl_SSL_Object *self, PyObject
+        // Free resources that can be freed according to OpenSSL docs
+        BN_free(x);
+        BN_free(y);
+        EC_KEY_free(ec);
```
Nice

nabla-c0d3 commented on tests/openssl_server/__init__.py:
```diff
@@ -105,7 +107,8 @@ def get_verify_argument(cls, client_auth_config: ClientAuthConfigEnum) -> str:
             self,
             client_auth_config: ClientAuthConfigEnum = ClientAuthConfigEnum.DISABLED,
-            extra_openssl_args: str = ''
+            extra_openssl_args: Optional[List[str]] = [],
```
this is a List[str]  (not Optional, which means None would be allowed)

nabla-c0d3 commented on tests/openssl_server/__init__.py:
```diff
@@ -177,6 +184,20 @@ class LegacyOpenSslServer(_OpenSslServer):
+            self,
+            client_auth_config: ClientAuthConfigEnum = ClientAuthConfigEnum.DISABLED,
+            cipher: Optional[str] = None,
+            prefer_server_order: Optional[bool] = False
```
This is a bool (not Optional)

nabla-c0d3 commented on tests/ssl_client_test.py:
```diff
@@ -160,6 +160,87 @@ def test_get_verified_chain_but_validation_failed(self):
+            assert 'size' in dh_info.keys()
+            assert 'x' in dh_info.keys()
+            assert 'y' in dh_info.keys()
+            assert 'public_key' in dh_info.keys()
```
Very cool

nabla-c0d3 reviewed (changes requested):
Hello, this is a very cool contribution :). Thanks a lot! I left some feedback but if you don't have time to address it, no worries - I'll take care of it myself when I look at this one more time (hopefully soon).
Thanks again!

FWinterborn commented:
Thanks for looking at this and for the feedback. I'll make those changes and update the branch.

fwinterbornbb committed:
Changed get_dh_info return type to dataclass
Return byte arrays rather than hex strings from native lib
Removed incorrect Optional type hints

nabla-c0d3 commented:
Thanks for making the changes - this is a great contribution :)

nabla-c0d3 commented:
Just in case you're interested I made some minor changes after merging: https://github.com/nabla-c0d3/nassl/commit/b5b65599cc3a17806d00728953e744d775d09081
Thanks again!
